File: libs/player.c
#include "player.h"
#include <stdlib.h>

static int rating[2][2] = {{1, 5}, {0, 3}};

static int compare_players(const void *vp1, const void *vp2)
{
	player *p1 = (player *)vp1, *p2 = (player *)vp2;
	return p2->score - p1->score;
}

static void sort_players(player *players, int count)
{
	qsort(players, count, sizeof(player), compare_players);
}

void clear_score(player *players, int count)
{
	for (int i = 0; i < count; ++i)
	{
		players[i].score = 0;
	}
}

void play_game(player *players, int count, int loops)
{
	bool *answers1 = (bool *)malloc(sizeof(bool) * loops);
	bool *answers2 = (bool *)malloc(sizeof(bool) * loops);
	for (int i = 0; i < count; ++i)
	{
		for (int j = 0; j < count; ++j)
		{
			for (int k = 0; k < loops; ++k)
			{
				answers1[k] = players[i].strategy(answers2, k);
				answers2[k] = players[j].strategy(answers1, k);
				players[i].score += rating[answers1[k]][answers2[k]];
			}
		}
	}
	free(answers1);
	free(answers2);

	sort_players(players, count);
}

File: src/player.rs
use crate::strategy;

pub struct Player {
    strategy: Box<dyn strategy::Strategy>,
    points: Vec<i32>,
    i: usize
}

impl Player {
    pub fn name(&self) -> String {
        format!("{:?}", self.strategy)
    }

    pub fn description(&self) -> String {
        self.strategy.description()
    }

    pub fn score<'a>(&'a self) -> &'a [i32] {
        &self.points
    }

    pub fn index(&self) -> usize {
        self.i
    }
}

pub fn all_players() -> Vec<Player> {
    strategy::all_strategies().into_iter().enumerate().map(|(i, strategy)| {
        Player { strategy, points: Vec::new(), i }
    }).collect()
}

pub fn play_game(mut players: Vec<Player>, rounds: usize, rating: &[[i32; 2]]) -> Vec<Player> {
    let mut answers1 = vec![false; rounds];
    let mut answers2 = vec![false; rounds];

    for i in 0..players.len() {
        players[i].points = vec![0; rounds];

        for j in 0..players.len() {
            for k in 0..rounds {
                answers1[k] = players[i].strategy.play(&answers2[0..k]);
                answers2[k] = players[j].strategy.play(&answers1[0..k]);
                players[i].points[k] += rating[answers1[k] as usize][answers2[k] as usize];
            }
        }

        for k in 1..rounds {
            players[i].points[k] += players[i].points[k-1];
        }
    }

    players.sort_by_key(|p| -p.points[rounds-1]);
    players
}
